### PR TITLE
fix(image): fix pixelDensity idempotence and high-DPI canvas resize

### LINKED ISSUE #9152: [p5.js 2.0+ Bug Report]: p5.Image pixelDensity() compounds division on repeated calls and resize() desynchronizes canvas backing store
### Most appropriate sub-area of p5.js?

- [ ] Accessibility
- [ ] Color
- [ ] Core/Environment/Rendering
- [ ] Data
- [ ] DOM
- [ ] Events
- [x] Image
- [ ] IO
- [ ] Math
- [ ] Typography
- [ ] Utilities
- [ ] WebGL
- [ ] WebGPU
- [ ] p5.strands
- [ ] Build process
- [ ] Unit testing
- [ ] Internationalization
- [ ] Friendly errors
- [ ] Other (specify if possible)

### p5.js version

2.x (main branch)

### Web browser and version

All browsers (Chromium / Firefox / Safari)

### Operating system

macOS / All

### Steps:

1. Create a `p5.Image` (e.g. `const img = createImage(100, 100)`).
2. Set `img.pixelDensity(2)` — `img.width` becomes `50`.
3. Call `img.pixelDensity(2)` again — `img.width` becomes `25` instead of remaining `50`.
4. Call `img.pixelDensity(1)` — `img.width` remains `25` instead of restoring to `100`.
5. Call `img.resize(50, 50)` on an image with `pixelDensity(2)` — `img.canvas.width` is set to `50` instead of `100` (`50 * 2`), causing `img.get(30, 30)` to return `[0, 0, 0, 0]` as out of bounds, and `img.set(x, y)` to silently drop writes.

### Snippet:

```js
function setup() {
  createCanvas(200, 200);

  // --- Bug 1: pixelDensity setter is not idempotent & cannot be reset ---
  const img1 = createImage(100, 100);
  img1.pixelDensity(2);
  console.log('After pixelDensity(2):', img1.width); // 50 (correct)

  img1.pixelDensity(2);
  console.log('After 2nd pixelDensity(2):', img1.width); // 25 (BUG: expected 50, halved again)

  img1.pixelDensity(1);
  console.log('After resetting pixelDensity(1):', img1.width); // 25 (BUG: expected 100, not restored)

  // --- Bug 2: resize() desynchronizes canvas backing store at density > 1 ---
  const img2 = createImage(100, 100);
  img2.pixelDensity(2);
  img2.resize(50, 50);

  console.log('Logical width:', img2.width); // 50
  console.log('Backing canvas.width:', img2.canvas.width); // 50 (BUG: expected 100)

  // Because canvas.width is 50, get() checks x * 2 >= canvas.width.
  // Coordinate (30, 30) is inside the 50x50 logical image, but 30 * 2 = 60 >= 50.
  const col = img2.get(30, 30);
  console.log('img2.get(30, 30):', col); // [0, 0, 0, 0] (BUG: treated as out of bounds)
}
```

### Expected behavior:

1. `img.pixelDensity(density)` should be idempotent: calling `img.pixelDensity(2)` repeatedly should preserve logical dimensions at `canvas.width / density` (`50`), and calling `img.pixelDensity(1)` should restore logical dimensions to `canvas.width / 1` (`100`).
2. `img.resize(width, height)` should maintain high-DPI resolution by resizing the backing canvas to `Math.floor(width * this._pixelDensity)` and `Math.floor(height * this._pixelDensity)`. Pixel operations (`get()`, `set()`, `loadPixels()`) should work accurately across the full logical dimensions of the resized image.

### Actual behavior:

1. In `src/image/p5.Image.js`:
```js
// Adjust canvas dimensions based on pixel density
this.width /= density;
this.height /= density;
```
`this.width` and `this.height` are mutated via relative division (`/= density`). Each call compounds the division, and resetting to `1` divides by `1` without restoring the original dimensions.

2. In `src/image/p5.Image.js`:
```js
// Resize the original canvas, which will clear its contents
this.canvas.width = this.width = width;
this.canvas.height = this.height = height;
```
`resize()` sets `this.canvas.width` and `this.canvas.height` directly to the logical target size `width` and `height`, discarding physical pixel density and desynchronizing the backing canvas from `this._pixelDensity`. As a result, subsequent `get(x, y)` calls fail bounds checking for any coordinates where `x * pd >= canvas.width`, returning transparent black `[0, 0, 0, 0]`, and `set(x, y)` silently drops pixel writes.

### PR DESCRIPTION
**Pull request by Pcmhacker-piro**
Resolves #9152

### Changes:
- In `src/image/p5.Image.js`, fixed `pixelDensity(density)` to derive logical dimensions from physical canvas dimensions (`this.width = this.canvas.width / density`, `this.height = this.canvas.height / density`). This makes repeated calls idempotent (no compounding dimension shrinkage) and allows resetting `pixelDensity(1)` to accurately restore original dimensions.
- In `src/image/p5.Image.js`, fixed `resize(width, height)` to scale the backing canvas and temporary canvas by `this._pixelDensity` (`width * pd`, `height * pd`). This prevents high-DPI images from having their canvas backing store downscaled to logical size, which caused subsequent `get()` calls to return `[0, 0, 0, 0]` as out-of-bounds and `set()` to fail writes.
- Used duck typing (`imgOrCol?.isColor`) in `Image.prototype.set()` to avoid cross-module `Color` imports and prevent `ReferenceError: p5 is not defined`.
- Added unit tests in `test/unit/image/p5.Image.js` covering:
  - `pixelDensity()` getter and setter behavior
  - Repeated `pixelDensity()` idempotence and restoring dimensions
  - Non-positive density fallback to 1
  - Backing canvas scaling with `pixelDensity > 1` in `resize()`
  - Full coordinate access with `get()` and `set()` after resizing a high-DPI image

### PR Checklist
- [x] `npm run lint` passes
- [ ] [Inline reference] is included / updated
- [x] [Unit tests] are included / updated

### AI Usage Disclosure
AI tools were used assistively to help write unit test cases and verify edge cases. The changes have been tested and reviewed locally.

### CODE REVIEW

**Comment by Pcmhacker-piro** [comment]
Hi @ksen0, could you please take a look at this PR when you have a moment? It resolves #9152 by ensuring `p5.Image.prototype.pixelDensity` idempotence and properly scaling the backing canvas during `resize()`. Comprehensive unit tests have been added, and both tests and linter are passing. Thank you!

File: src/image/p5.Image.js [hunk 1]
```diff
@@ -611,7 +612,7 @@ class Image {
           a = imgOrCol[3];
           //this.updatePixels.call(this);
         }
-      } else if (imgOrCol instanceof p5.Color) {
```
Comment (davepagurek) [inline src/image/p5.Image.js:614]: @limzykenneth I wanted to get your thoughts on this: if we want to have the ability to load all modules as separate files in the future then we may want to use `p5.` prefixes more to avoid imports, but if we're thinking of implementing different combinations of modules as separate single-file builds then switching to imports like this makes sense. It's a little confusing if we have both, e.g. for WebGPU where that feels like it should be a separate file, because it means using a different convention in one spot than another, but maybe that's ok?

**Review by davepagurek — APPROVED** [review]
The core changes look good to me! Just left one thing I wanted to clarify about how we generally want to reference other modules within the code before merging.

File: src/image/p5.Image.js (hunk 1 above)
Comment (Pcmhacker-piro) [reply src/image/p5.Image.js:614]: Thanks @davepagurek! To avoid introducing a cross-module import of `Color` while also fixing the issue where referencing `p5.Color` threw `ReferenceError: p5 is not defined` (since `p5` is not in module scope within `p5.Image.js`), I've updated this to use duck typing (`imgOrCol?.isColor`). This matches how `Color` is checked in `p5.Shader` and `RendererWebGPU` (leveraging `isColor = true` from `p5.Color.js`), removing the import and keeping separate module/bundle builds clean. Pushed the update in ebdb78c!

**Comment by Pcmhacker-piro** [comment]
Hi @davepagurek, I have updated the PR to address your feedback:

- Removed the cross-module `import { Color } from '../color/p5.Color'` from `src/image/p5.Image.js`.
- Switched to duck typing (`imgOrCol?.isColor`) in `Image.prototype.set()`, matching the established convention in `p5.Shader` and `RendererWebGPU` (backed by `isColor = true` in `p5.Color.js`). This avoids cross-module imports while also avoiding `ReferenceError: p5 is not defined` (since `p5` is not in module scope within `p5.Image.js`).
- Verified all unit tests and `npm run lint` pass locally.

Ready for final review whenever you have a chance! Thank you!

**Review by davepagurek — APPROVED** [review]
looks good, thanks!

**Comment by davepagurek** [comment]
@all-contributors please add @Pcmhacker-hero for code, bug

**Comment by Pcmhacker-piro** [comment]
Hi  @davepagurek  

Since my PR has been merged, could you please add me to the contributor list as well? Thank you.

**Comment by Pcmhacker-piro** [comment]
Hi @davepagurek @perminder-17,

Thank you for merging!

In the bot command above, my other account `@Pcmhacker-hero` was incorrectly added instead of this `@Pcmhacker-piro` account, which resulted in PR #9179.

Please add my **@Pcmhacker-piro** account to the contributors list instead. All 3 of my pull requests in this repo are contributed from this account:
- #9167 (*merged*)
- #9170 (*merged*)
- #9168 (*open*)

Could you please close PR #9179 and run the following command to add **@Pcmhacker-piro**?

> @all-contributors please add @Pcmhacker-piro for code, test, bug

Thank you!
